runtime/vm/JFRChunkWriter.cpp

```diff
@@ -962,4 +962,261 @@ VM_JFRChunkWriter::writeThreadStatisticsEvent(void *anElement, void *userData)
+		cursor += snprintf(cursor, resultBufferEnd - cursor, "blocked on: ");
+	} else if ((J9VMTHREAD_STATE_WAITING == state) || (J9VMTHREAD_STATE_WAITING_TIMED == state)) {
+		cursor += snprintf(cursor, resultBufferEnd - cursor, "waiting on: ");
+	} else if ((J9VMTHREAD_STATE_PARKED == state) || (J9VMTHREAD_STATE_PARKED_TIMED == state)) {
+		cursor += snprintf(cursor, resultBufferEnd - cursor, "parked on: ");
+	} else {
+		cursor += snprintf(cursor, resultBufferEnd - cursor, "\n");
+		return cursor;
+	}
+
+	if (NULL != monitorObject) {
+		cursor = writeObject(vm, monitorObject, cursor, resultBufferEnd);
+	} else {
+		cursor += snprintf(cursor, resultBufferEnd - cursor, "<unknown>");
+	}
+	cursor += snprintf(cursor, resultBufferEnd - cursor, "\n");
+
+	return cursor;
+}
+
+static char *
+writeStacktrace(J9VMThread *currentThread, J9VMThread *walkThread, char *cursor, const char *resultBufferEnd)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9StackWalkState stackWalkState = {0};
+	const int maxMonitorInfosPerThread = 32;
+	J9ObjectMonitorInfo monitorInfos[maxMonitorInfosPerThread];
+	memset(monitorInfos, 0, maxMonitorInfosPerThread * sizeof(J9ObjectMonitorInfo));
+
+	IDATA monitorCount = vmFuncs->getOwnedObjectMonitors(currentThread, walkThread, monitorInfos, maxMonitorInfosPerThread, FALSE);
+
+	stackWalkState.walkThread = walkThread;
+	stackWalkState.flags =
+		J9_STACKWALK_ITERATE_FRAMES
+		| J9_STACKWALK_INCLUDE_NATIVES
+		| J9_STACKWALK_VISIBLE_ONLY
+		| J9_STACKWALK_RECORD_BYTECODE_PC_OFFSET
+		| J9_STACKWALK_NO_ERROR_REPORT;
+	stackWalkState.skipCount = 0;
+	stackWalkState.frameWalkFunction = stackWalkCallback;
+	stackWalkState.userData1 = cursor;
+	stackWalkState.userData2 = (void *)resultBufferEnd;
+	stackWalkState.userData3 = monitorInfos;
+	stackWalkState.userData4 = (void *)monitorCount;
+
+	vmFuncs->haltThreadForInspection(currentThread, walkThread);
+	vm->walkStackFrames(currentThread, &stackWalkState);
+	vmFuncs->resumeThreadForInspection(currentThread, walkThread);
+
+	cursor = (char *)stackWalkState.userData1;
+	cursor += snprintf(cursor, resultBufferEnd - cursor, "\n");
+	return cursor;
+}
+
+U_8 *
+VM_JFRChunkWriter::writeThreadDumpEvent()
+{
+	/* reserve size field */
+	U_8 *dataStart = _bufferWriter->getAndIncCursor(sizeof(U_32));
+
+	_bufferWriter->writeLEB128(ThreadDumpID);
+
+	/* write start time */
+	_bufferWriter->writeLEB128(j9time_current_time_millis());
+
+	char *resultBuffer = (char *)j9mem_allocate_memory(sizeof(char) * THREAD_DUMP_EVENT_SIZE, OMRMEM_CATEGORY_VM);
+	const char *resultBufferEnd = resultBuffer + THREAD_DUMP_EVENT_SIZE;
+
+	if (NULL != resultBuffer) {
+		J9VMThread *walkThread = J9_LINKED_LIST_START_DO(_vm->mainThread);
+		UDATA numThreads = 0;
+		char *cursor = resultBuffer;
+		J9InternalVMFunctions *vmFuncs = _vm->internalVMFunctions;
+		bool acquiredVMAccess = false;
+
+		if (0 == (_currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS)) {
```

Comment:
we should always have VMAccess, you can assert with `Assert_VM_mustHaveVMAccess`